swap/swap.go

```diff
@@ -434,6 +451,19 @@ func (s *Swap) Balances() (map[enode.ID]int64, error) {
 	return balances, nil
 }
 
+// AvailableBalance returns the total balance of the chequebook against which new cheques can be written
+func (s *Swap) AvailableBalance() (uint64, error) {
```

Comment:
I would really recommend avoiding 2 blockchain wide event filters in one function, this can take quite a while on a real network (especially if using remote endpoints like infura). Instead we could compute this per peer using `s.contract.availableBalanceFor(peer) + s.paidOut(p.beneficiary) - p.getLastSentCheque().CumulativeAmount`.

If we need this for all peers (not sure why we would though) we can compute `liquidBalance() + SUM(s.paidOut(peer)) - SUM(p.getLastSentCheque().CumulativeAmount)`. The last sum can be computed locally since we have all cheques in our store.

If we implement it that way we also don't need `s.paidOut`.

Reply:
Hi Ralph, this call is needed for a node operator to figure out how much balance is still left in his chequebook against which new cheques can be written. It allows him to figure out what his spendable balance is and when he has to re-deposit. Furthermore, this call is needed by Swarm in order to verify that an to-be-send cheque is not overspending.

I don't think we can prevent doing an event-filter on the smart-contract, but I can make it in such a way that we only do it upon startup.

I don't understand your proposed (second) formula, as I don't think they yield the correct result.

Currently `s.paidOut()` is actually `SUM(p.getLastSentCheque().CumulativeAmount)`
Do you mean that `s.paidOut(peer)` is the sum of cheques cashed in the smart-contract? If so, your formula would be correct, but we still need to filter for `chequeCashed` events.

Reply:
Hi Ralph, 
Based on your comment, I removed `paidOut` from the `swap` structure and place `availableBalance` there instead. This change minimizes the need to ever call the `ComputeAvailableBalance` function to one (boot-up).

If we would go for implementing your (second) formula instead of how it is implemented now with the event filters, we need to do a call to the smart-contract to read paidOut for every peer.

Is this indeed quicker than the event filter? Note that Infura recently updated how they handle event queries: https://blog.infura.io/faster-logs-and-events-e43e2fa13773

If you think that what you propose is indeed much quicker, I will make the update!

Reply:
Note: the option you presented and the one which is currently implemented are not entirely the same. They deviate in the case where a cheque is sent outside of Swarm and already cashed. The current implementation would recognize this (and calculate the available balance correctly from the point of cashing onwards), while your proposal would never recognize this cheque as it is not cashed by a peer, hence the `availableBalance` will always be off.

Reply:
We should not consider the case with cheques outside of Swarm, this is always a user error and if done other things will break anyway. If we want the chequebook to be able to be used for other purposes we should introduce an RPC call for that so that the actual cheque creation and sending still takes place in the Swarm node.

Blockchain-wide event queries (from block 0 to "latest") can be really slow in my experience (Have you tested this call against ropsten or mainnet?). The Infura optimisation doesn't help much if it is then still slow for other users. (Since it's only being called once now it's not as big of a problem anymore).

While true that my formula does require more calls to `paidOut()`, this is an `eth_call` operation against the latest state which is a very fast operation. My per-peer formula would also have the advantage that it correctly accounts for any set hard deposit.

The current approach also doesn't update `availableBalance` correctly if the balance increases due to a cashed cheque after the node has started.

Reply:
> Have you tested this call against ropsten or mainnet?)

It is in the order of milliseconds. 

> My per-peer formula would also have the advantage that it correctly accounts for any set hard deposit.

Sure, we can add this also to the approach which I propose. 

> The current approach also doesn't update availableBalance correctly if the balance increases due to a cashed cheque after the node has started.

You mean that right now, we only take into account a deposit event when it is initiated by the node, and not when it is caused by a `cashCheque` call, with the beneficiary being the nodes contract address? Right now, the beneficiary of a cheque is always `p.beneficiary`, so `cashCheque` should not influence the `availableBalance`.

Reply:
While I see that your approach should be faster (theoretically), I do have the feeling that it is more fragile and I mentioned at least one edge case in which the approach with event filters would catch. Sure, when a cheque is been sent outside of the Swarm content, bad things can happen anyway, but I would still feel more comfortable by implementing something that would catch at least one edge case, than implementing something that does not catch this edge case but is somewhat faster. Especially, if the call only happens at boot-up as the case now. 

Please let me know what you think about this.

Reply:
> You mean that right now, we only take into account a deposit event when it is initiated by the node, and not when it is caused by a `cashCheque` call, with the beneficiary being the nodes contract address? Right now, the beneficiary of a cheque is always `p.beneficiary`, so `cashCheque` should not influence the `availableBalance`.

That's the beneficiary in the cheque but in our call to `CashChequeBeneficiary` we specify our chequebook as recipient, therefore it's the chequebooks balance that will increase. You would have to also install an event subscription (using `WatchDeposit`) and adjust `availableBalance` accordingly. The same goes for `Withdraw` if we allow withdraws outside of Swarm.

> While I see that your approach should be faster (theoretically), I do have the feeling that it is more fragile and I mentioned at least one edge case in which the approach with event filters would catch. Sure, when a cheque is been sent outside of the Swarm content, bad things can happen anyway, but I would still feel more comfortable by implementing something that would catch at least one edge case, than implementing something that does not catch this edge case but is somewhat faster. Especially, if the call only happens at boot-up as the case now.

I think the approach with reading the events only once is a lot more fragile to (unavoidable) edge cases. Especially once we also track deposits during while Swarm runs. Some scenarios that will be hard to handle with this approach:
* A deposit that happened right before the node started and then got reorged. We might either account it twice (if we have the watch and it gets reincluded) or we account it even though it never happened in retrospect (e.g. cause in the new chain another cheque came in first and we bounced).
* (if we have the watch and I don't think we can avoid it) we also might have problems on unstable internet connections. In my experience event subscriptions rarely survive through connection issues with the backend (especially if you might end up with another node at the endpoint like in Infura). Therefore I would try to avoid this in long running processes. Alternatively we could do regular polling for ranges of blocks but that would reintroduce the reorg problematic from the previous point.
* Once we want to account for hard deposits your formula will become more complicated, especially if hard deposits change while the node is online.

My approach wouldn't face any of these issues because it always gets the latest blockchain data when needed.

Reply:
Great answer Ralph. I am almost convinced :). Are you proposing to **always** do the call to availableBalance (not only during start-up), meaning we loop over the peers and do the required `eth_call`? 
Note that such a call has to happen frequently if we don't want a node to send "bad cheques".

Reply:
Updated.

Reply:
> Great answer Ralph. I am almost convinced :). Are you proposing to **always** do the call to availableBalance (not only during start-up), meaning we loop over the peers and do the required `eth_call`?

I would compute `availableBalance` on a per-peer basis whenever we want to send a cheque. Then it only requires one `eth_call` for every cheque sent.

If we want this `availableBalance` for all peers we have to do an `eth_call` for every peer but that would only happen when the RPC call is made and not during regular operation.

Reply:
great discussion guys!

i don't understand the issue as in-depth as you two, but at a high level, it seems sensible to me to query the source of the data (in this case the blockchain) for the latest info when it's needed.

but this, of course, is contingent upon these queries not slowing down the system significantly. it might be hard to verify this. 

is the alternative (proposed by @Eknir) slower overall? or is it just less robust/more complicated? i understand the query from the genesis block to the present is slow according to @ralph-pichler, but it would be less queries in the long run, right?

Reply:
@Eknir  we need `availableBalance` for all peers in order to know how far away we are from needing to make a new deposit, right?

Reply:
If I may put my thoughts here. I pretty much have to agree with @ralph-pichler.

> Blockchain-wide event queries (from block 0 to "latest") can be really slow in my experience (Have you tested this call against ropsten or mainnet?). The Infura optimisation doesn't help much if it is then still slow for other users. (Since it's only being called once now it's not as big of a problem anymore).

As @ralph-pichler says these events are horribly slow. We've been using them in Giveth and at one point it took 1.5hs to reprocess everything (around 5000 events). And this is after optimizations like querying since the contract deployment etc. Sure here it will not be that extreme in most of the cases but we should at minimum test it. And yeah chain reorgs or connection issues created a lot of errors and had to be handled.

> We should not consider the case with cheques outside of Swarm, this is always a user error and if done other things will break anyway.

I also agree here. At least for the immediate future, I don't think we should be too bothered by this case. We should make an issue about this but not actively solve it now.

One of the usecases I can imagine is when user runs multiple swarm nodes with one chequebook. Still this brings a lot of problems to solve on the accounting side because it introduces all sorts of concurrency issues.